chain/chain/src/chain.rs

```diff
@@ -2542,8 +2542,46 @@ impl Chain {
         &mut self,
         block_hash: &CryptoHash,
         shard_id: ShardId,
-    ) -> Result<Option<&CryptoHash>, Error> {
-        self.store.get_next_block_hash_with_new_chunk(block_hash, shard_id)
+    ) -> Result<Option<(CryptoHash, ShardId)>, Error> {
+        let mut block_hash = block_hash.clone();
+        let mut epoch_id = self.get_block_header(&block_hash)?.epoch_id().clone();
+        let mut shard_layout = self.runtime_adapter.get_shard_layout(&epoch_id)?;
+        let mut shard_ids = vec![shard_id];
+
+        while let Ok(next_block_hash) = self.store.get_next_block_hash(&block_hash) {
+            let next_block_hash = next_block_hash.clone();
+            let next_epoch_id = self.get_block_header(&next_block_hash)?.epoch_id().clone();
+            if next_epoch_id != epoch_id {
+                let next_shard_layout = self.runtime_adapter.get_shard_layout(&next_epoch_id)?;
+                if next_shard_layout != shard_layout {
+                    shard_ids = shard_ids
+                        .into_iter()
+                        .flat_map(|id| {
+                            next_shard_layout.get_split_shards(id).unwrap_or_else(|| {
+                                panic!("invalid shard layout {:?} because it does not contain split shards for parent shard {}", next_shard_layout, id)
+                            })
+                        })
+                        .collect::<Vec<_>>()
```

Comment:
nit: can this cause duplicates in the vector ? 

(only if we allow shards to be merged - not sure if we do).

Reply:
Yes, in the case of shards merging, this won't work. In fact, I didn't consider shard merging when I designed the shard layout API (you can tell from the fact that the function is called `split_shards` instead of something else.

Reply:
Also, right now, in the constructor, we check that there is no duplicates in the map. https://github.com/near/nearcore/blob/master/core/primitives/src/shard_layout.rs#L68